flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/collect/CollectResultFetcher.java

```diff
@@ -0,0 +1,343 @@
+			byte[] serializedResult = serializedResults.get(0);
+			return CollectSinkFunction.deserializeAccumulatorResult(serializedResult);
+		} catch (ClassNotFoundException | IOException e) {
+			// this is impossible
+			throw new RuntimeException("Failed to deserialize accumulator results", e);
+		}
+	}
+
+	private boolean isJobTerminated() {
+		checkJobClientConfigured();
+
+		try {
+			JobStatus status = jobClient.getJobStatus().get();
+			return status.isGloballyTerminalState();
+		} catch (Exception e) {
+			// TODO
+			//  This is sort of hack.
+			//  Currently different execution environment will have different behaviors
+			//  when fetching a finished job status.
+			//  For example, standalone session cluster will return a normal FINISHED,
+			//  while mini cluster will throw IllegalStateException,
+			//  and yarn per job will throw ApplicationNotFoundException.
+			//  We have to assume that job has finished in this case.
+			//  Change this when these behaviors are unified.
+			LOG.warn("Failed to get job status so we assume that the job has terminated. Some data might be lost.", e);
+			return true;
+		}
+	}
+
+	private void cancelJob() {
+		checkJobClientConfigured();
+
+		if (!isJobTerminated()) {
+			jobClient.cancel();
+		}
+	}
+
+	private void sleepBeforeRetry() {
+		if (retryMillis <= 0) {
+			return;
+		}
+
+		try {
+			// TODO a more proper retry strategy?
+			Thread.sleep(retryMillis);
+		} catch (InterruptedException e) {
+			LOG.warn("Interrupted when sleeping before a retry", e);
+		}
+	}
+
+	private void checkJobClientConfigured() {
+		Preconditions.checkNotNull(jobClient, "Job client must be configured before first use.");
+	}
+
+	private class ResultBuffer {
+
+		private static final String INIT_VERSION = "";
+
+		private final LinkedList<T> buffer;
+		private final TypeSerializer<T> serializer;
+
+		private String version;
+		private long offset;
+		private long lastCheckpointedOffset;
+		private long userHead;
+		private long userTail;
+
+		private ResultBuffer(TypeSerializer<T> serializer) {
+			this.buffer = new LinkedList<>();
+			this.serializer = serializer;
+
+			this.version = INIT_VERSION;
+			this.offset = 0;
+			this.lastCheckpointedOffset = 0;
+			this.userHead = 0;
+			this.userTail = 0;
+		}
+
+		private T next() {
+			if (userHead == userTail) {
+				return null;
+			}
+			T ret = buffer.removeFirst();
+			userHead++;
+
+			sanityCheck();
+			return ret;
+		}
+
+		private void dealWithResponse(CollectCoordinationResponse<T> response) {
+			dealWithResponse(response, offset);
+		}
+
+		private void dealWithResponse(CollectCoordinationResponse<T> response, long responseOffset) {
+			String responseVersion = response.getVersion();
+			long responseLastCheckpointedOffset = response.getLastCheckpointedOffset();
+			List<T> results;
+			try {
+				results = response.getResults(serializer);
+			} catch (IOException e) {
```

Comment:
I don't think we should eat this exception, we should throw this out to show something is going wrong